pppd/pppd.h

```diff
@@ -57,7 +57,18 @@
 #include <sys/param.h>		/* for MAXPATHLEN and BSD4_4, if defined */
 #include <sys/types.h>		/* for u_int32_t, if defined */
 #include <sys/time.h>		/* for struct timeval */
+
+/*
+ * linux kernel does not have net/ppp_defs.h by default.
+ * this fix is for musl as it does not create net/ppp_defs.h
+ * like glibc.
+*/
+
+#ifdef __linux__
```

Comment:
How about:
```
#ifdef HAVE_LINUX_PPP_DEFS_H
#include <linux/ppp_defs.h>
#endif

#ifdef HAVE_NET_PPP_DEFS_H
#include <net/ppp_defs.h>
#endif
```

Then ideally, the logic of setting the defines would be in pppd/Makefile.linux, this would be something similar to how it currently detects utmp.h.

```
# Use libutil; test if logwtmp is declared in <utmp.h> to detect
UTMPHEADER = "\#include <utmp.h>"
ifeq ($(shell echo $(UTMPHEADER) | $(CC) -E - 2>/dev/null | grep -q logwtmp && echo yes),yes)
USE_LIBUTIL=y
endif
```

Now swap that out for linux/ppp_defs.h, or net/ppp_defs.h

It would be easy if the autotools change list would be merge. Your update to the pppd.h would still remain, but Makefile.linux section would be replace with an AC_CHECK_HEADER.

Reply:
The ifdef you suggest can't work as glibc has both linux/ and net/ so it would check for both. Personally I don't know about uclibc

Reply:
But there should be a combination that would work for all, e.g. if have <net/ppp_defs.h> use that, or use <linux/ppp_defs.h>, vice versa

Reply:
Also, to @paulusmack point, why is it that gentoo/muslc does not provide this file, and buildroot does? Did you configure the muslc library correctly, or are you missing a patch?

Reply:
The net/ppp_defs.h is not provided by kernel but libc itself. There's a patch like this one http://lists.busybox.net/pipermail/buildroot/2016-March/156823.html for it

Reply:
What are you trying to build, pppd, or some third part plugin? The pppd can (and should) use the provided ppp_defs.h if the system has it missing

But if you are looking at building a ppp plugin, then you may not need this. The pppd/*.h headers is a bit of a cluster f** and needs some TLC.

Reply:
PPP itself build without problems but things like NetworkManager does not build because missing net/ppp_defs.h. when I tried changing #include <net/ppp_defs.h> to linux/ works without problems

Reply:
I see, which part of network manager are you trying to compile? 

Ideally, I think the Network Manager project should include the right ppp_defs.h and use meson, or autoconf to setup the correct include prior to including ppp/pppd.h for the system it is being compiled on. 

As I am starting to think of this, the pppd project should drop the include from pppd.h and require others to fix that up. The pppd project could wrap it under #ifder/#endif stuff, but it is contingent upon configuring their system correctly with defines dictated by pppd.

That sort of dependency i.e. network-manager needs to know what define to enable to compile the source with seems like a no go, i.e. wrap it with HAVE_NET_PPP_DEFS_H.  

I've been advocating for this header include mess cleanup for a while...

Reply:
I have an idea: Instead of using #ifdef __linux___ we could use #!defined __GLIBC__ or something similar, that's because uclibc follows glibc way while musl doesn't. Can it be a solution?

Reply:
@enaess: What do you think?

Reply:
I've tried with no luck